core/bessctl.cc

```diff
@@ -1376,6 +1463,107 @@ class BESSControlImpl final : public BESSControl::Service {
+    if (request->target_case() ==
+        ConfigureResumeHookRequest::TargetCase::kModule) {
+      const auto factory =
+          bess::ResumeHookFactory<Module>::all_resume_hook_factories().find(
+              request->hook_name());
+      if (factory ==
+          bess::ResumeHookFactory<Module>::all_resume_hook_factories().end()) {
+        return return_with_error(response, ENOENT, "No such resume hook: %s",
+                                 request->hook_name().c_str());
+      }
+
+      const auto& it = ModuleBuilder::all_modules().find(request->module());
+      if (it == ModuleBuilder::all_modules().end()) {
+        return return_with_error(response, ENOENT, "No module '%s' found",
+                                 request->module().c_str());
+      }
+
+      ResumeHookUser<Module>* user =
+          ResumeHookUser<Module>::get_user_hooks(it->second);
+      if (request->enable()) {
+        VLOG(1) << "Enabling resume hook '" << request->hook_name()
+                << "' for module '" << request->module() << "'";
+        *response = enable_resume_hook(user, factory->second, it->second,
+                                       request->arg());
+        if (response->has_error()) {
+          return Status::OK;
+        }
+        VLOG(1) << "Enabled resume hook '" << request->hook_name()
+                << "' for module '" << request->module() << "'";
+      } else {
+        disable_resume_hook(user, request->hook_name());
+        VLOG(1) << "Disabled resume hook '" << request->hook_name()
+                << "' for module '" << request->module() << "'";
+      }
+      return Status::OK;
+    }
+
+    const auto factory =
+        bess::ResumeHookFactory<Worker>::all_resume_hook_factories().find(
+            request->hook_name());
+    if (factory ==
+        bess::ResumeHookFactory<Worker>::all_resume_hook_factories().end()) {
+      return return_with_error(response, ENOENT, "No such resume hook: %s",
+                               request->hook_name().c_str());
+    }
+
+    int wid = 0;
+    ResumeHookUser<Worker>* user;
+    if (request->global()) {
+      user = ResumeHookUser<Worker>::get_user_hooks(nullptr);
+      if (request->enable()) {
+        VLOG(1) << "Enabling global resume hook '" << request->hook_name()
+                << "'";
+        *response = enable_resume_hook<Worker>(user, factory->second, nullptr,
+                                               request->arg());
+        if (response->has_error()) {
+          return Status::OK;
+        }
+        VLOG(1) << "Enabled global resume hook '" << request->hook_name()
+                << "'";
+      } else {
+        disable_resume_hook(user, request->hook_name());
+        VLOG(1) << "Disabled global resume hook '" << request->hook_name()
+                << "'";
+      }
+      return Status::OK;
+    }
+
+    uint64_t worker_filter = request->workers();
+    while (worker_filter) {
+      Worker* w = workers[wid];
```

Comment:
The documentation says that the filter is a bitmap, but we don't actually test the specific bit (`worker_filter & 1`) here...

Reply:
Doh. Thanks for catching-- will fix

Reply:
Whoops, the `continue` skips the `worker_filter >>= 1` and `wid++` lines 😮

Reply:
Oh, yikes! Fixed and rebased